platform/linuxbsd/os_linuxbsd.cpp

```diff
@@ -303,65 +303,150 @@ static String get_mountpoint(const String &p_path) {
 		}
 	}
 
 	// Issue an error if none of the previous locations is appropriate for the trash can.
-	if (trash_can == "") {
+	if (trash_path == "") {
 		ERR_PRINT("move_to_trash: Could not determine the trash can location");
 		return FAILED;
 	}
 
 	// Create needed directories for decided trash can location.
-	DirAccess *dir_access = DirAccess::create(DirAccess::ACCESS_FILESYSTEM);
-	Error err = dir_access->make_dir_recursive(trash_can);
-	memdelete(dir_access);
+	{
+		DirAccess *dir_access = DirAccess::create(DirAccess::ACCESS_FILESYSTEM);
+		Error err = dir_access->make_dir_recursive(trash_path);
+		Error err2 = dir_access->make_dir_recursive(trash_path + "/files");
+		Error err3 = dir_access->make_dir_recursive(trash_path + "/info");
+
+		// Issue an error if trash can is not created proprely.
+		if (err != OK || err2 != OK || err3 != OK) {
+			ERR_PRINT("move_to_trash: Could not create the trash can \"" + trash_path + "\"");
+			return err;
+		}
+		memdelete(dir_access);
+	}
 
-	// Issue an error if trash can is not created proprely.
-	if (err != OK) {
-		ERR_PRINT("move_to_trash: Could not create the trash can \"" + trash_can + "\"");
-		return err;
+	// The trash can is successfully created, now we check that we don't exceed our file name Length limit.
+	// If the file name it's too long trim it so we can add the identifyng number and ".trashinfo"
+	// Assumes that the file name Length limit is 255 characters.
+	String file_name = basename(p_path.utf8().get_data());
+	if (file_name.length() > 240) {
+		file_name = file_name.substr(0, file_name.length() - 15);
+	}
+
+	String dest_path = trash_path + "/files/" + file_name;
+	struct stat buff;
+	int id_number = 0;
+	String fn = file_name;
+
+	// Checks if a resource with the same name already exist in the trash can,
+	// if there is, add an identifying number to our resource's name.
+	while (stat(dest_path.utf8().get_data(), &buff) == 0) {
+		id_number++;
+
+		// Added a limit to check for identically named files already on the trash can
+		// if there are too many it could make the editor unresponsive.
+		if (id_number > 999) {
+			ERR_PRINT("move_to_trash: Too many identically named resources already in the trash can");
+			return FAILED;
+		}
+		fn = file_name + "." + itos(id_number);
+		dest_path = trash_path + "/files/" + fn;
+	}
+	file_name = fn;
+
+	// Generates the .trashinfo file
+	OS::Date date = OS::get_singleton()->get_date(false);
+	OS::Time time = OS::get_singleton()->get_time(false);
+	String timestamp = vformat("%04d-%02d-%02dT%02d:%02d:", date.year, date.month, date.day, time.hour, time.min);
+	timestamp = vformat("%s%02d", timestamp, time.sec); // vformat only supports up to 6 arguments.
+	String trash_info = "[Trash Info]\nPath=" + p_path.http_escape() + "\nDeletionDate=" + timestamp + "\n";
+	{
+		Error err;
+		FileAccess *file = FileAccess::open(trash_path + "/info/" + file_name + ".trashinfo", FileAccess::WRITE, &err);
+		if (err != OK) {
```

Comment:
Same thing with ERR_FAIL_COND_V_MSG here.

```suggestion
		ERR_FAIL_COND_V_MSG(err != OK, err, "Can't create trashinfo file:" + trash_path + "/info/" + file_name + ".trashinfo");
```

Reply:
Done and done.